AX: The accessibility tree doesn't update when selection is cleared

twilco opened the pull request:
#### c726e39c4422ee5635c91936773212032cc1c2e7
<pre>
AX: The accessibility tree doesn&apos;t update when selection is cleared
<a href="https://bugs.webkit.org/show_bug.cgi?id=279484">https://bugs.webkit.org/show_bug.cgi?id=279484</a>
<a href="https://rdar.apple.com/135768164">rdar://135768164</a>

Reviewed by Andres Gonzalez.

Prior to this patch, we never updated the isolated tree when selection
was cleared, only when it was changed from one node to another. This
meant that we would serve stale data for the AXSelectedTextMarkerRangeAttribute
API.

* LayoutTests/accessibility/mac/dynamic-selection-expected.txt: Added.
* LayoutTests/accessibility/mac/dynamic-selection.html: Added.
* LayoutTests/resources/accessibility-helper.js:
* Source/WebCore/editing/mac/FrameSelectionMac.mm:
(WebCore::FrameSelection::notifyAccessibilityForSelectionChange):

Canonical link: <a href="https://commits.webkit.org/283506@main">https://commits.webkit.org/283506@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/48dfb7deb6bbd3c7687d0ef4f91e94dd97eea783

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66454 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45829 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19075 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70487 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/17587 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/68572 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53628 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17347 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53297 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11895 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69521 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42235 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57523 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33951 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38906 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/14911 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15940 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60809 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15253 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72190 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10411 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/14625 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/60624 "Found 5 new test failures: imported/w3c/web-platform-tests/css/css-transforms/animation/scale-and-rotate-both-specified-on-animation-keyframes.html imported/w3c/web-platform-tests/css/css-transforms/animation/transform-interpolation-skew.html imported/w3c/web-platform-tests/css/css-transforms/animation/transform-interpolation-translate.html imported/w3c/web-platform-tests/css/css-transforms/animation/transform-percent-with-width-and-height-separate.html imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-animated-001.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10443 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57592 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60938 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/14681 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8593 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2203 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/41636 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/42713 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/43896 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42456 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

AndresGonzalezApple commented on LayoutTests/accessibility/mac/dynamic-selection.html:
```diff
@@ -0,0 +1,56 @@
+    var webArea = accessibilityController.rootElement.childAtIndex(0);
+    var selectedRange;
+    setTimeout(async function() {
+        selectedRange = await selectElementTextById("span-one");
```
Nit: should selectElementTextById take the webArea since it is available here instead of calculating it several times?

twilco replied:
Sure, can do that.

twilco replied:
Fixed in latest commit!

AndresGonzalezApple commented on LayoutTests/accessibility/mac/dynamic-selection.html:
```diff
@@ -0,0 +1,56 @@
+}
+</script>
+</body>
+</html>
```
This test is fine and it follows a pattern we have implemented in many tests. A more realistic test in the sense of mimicking AT interaction more closely, we should consider moving towards an event notification driven model. This is that instead of waitFor, use a listener for the selection changed notification and query for the new notification in the listener. That's how AT operates and the test will check both the correct notifications and the property retrieval. Not saying that this test is wrong, but perhaps this is a good turning point to further adopt the event driven model in our tests.

twilco replied:
Good idea, let's take that strategy for future tests.

AndresGonzalezApple commented on LayoutTests/resources/accessibility-helper.js:
```diff
@@ -266,6 +266,29 @@ async function waitForFocus(id) {
+        return webArea.textMarkerRangeLength(selectedRange) > 0;
+    });
+    return selectedRange;
+}
```
This is duplicated.

twilco replied:
Can you elaborate on what is duplicated? This function operates in a few steps:

 1. Clear DOM selection entirely.
 2. Wait for the accessibility tree to report there is no selected text marker range.
 3. Select the text of the element by `id`.
 4. Wait for the accessibility tree to report there is a selected text marker range (its length is greater than zero). Because we cleared the selection in steps 1 and 2, we know the range we get here is of the text we meant to select.
 
If you have a better idea on doing this, let me know.

twilco replied:
Facepalm, Andres meant that this function is in `accessibility-helper.js` and in the test! Will fix this, thanks.

AndresGonzalezApple commented on Source/WebCore/editing/mac/FrameSelectionMac.mm:
```diff
@@ -40,9 +40,19 @@
+            // changes would be needed in VoiceOver too, so just do onSelectedTextChanged for now.
+#if ENABLE(ACCESSIBILITY_ISOLATED_TREE)
+            cache->onSelectedTextChanged(m_selection);
+#endif
```
VO doesn't need to announce that text is unselected, although it may do it in some cases. . But regardless of VO response in this case, we should fire the right notification.
Does this need to be iso tree specific? Shouldn't it work the same for all ports?

twilco replied:
> AG: VO doesn't need to announce that text is unselected, although it may do it in some cases.

TW: It announces "foo selected, bar unselected" when the selection changes, so I think reading "bar unselected" when bar is unselected without anything becoming selected at the same time could make sense and be useful for folks. Right now, the selection disappears without any announcement, which could maybe cause confusion for users.

> AG: But regardless of VO response in this case, we should fire the right notification.

TW: Agreed in general, but posting the right notification is a bit complicated:

1. VoiceOver expects the notification to be associated with an actual text object, and we don't have that information when text becomes unselected — the associated DOM node with an unselection is null. I tried posting from the root web area, but that does not work, and it's not clear that's even the right thing to do. So we would have to figure out how to pipe that information through, or start tracking the last selected object maybe?
2. It's not clear what `AXTextStateChangeType` unselection maps to — none are a good fit.

I personally don't think right now is the right time to fully unravel this thread, so I was hoping to make a bit of incremental progress and fix the ITM cache invalidation bug. But I could file a separate bug and add it here in a comment? What do you think? Do you think we should get all the way to the bottom here, or make a bug for follow-up and take the intermediate bug fix for now?

> AG: Does this need to be iso tree specific? Shouldn't it work the same for all ports?

TW: I'd say yes, port agnosticism would be better in general, however the definition of `AXObjectCache::onSelectedTextChanged` is gated by `PLATFORM(MAC)`. I could mention this in the follow-up bug, or happy to address in some other way if you have an idea!

twilco replied:
Andres and I spoke, and resolution is to file a separate bug with some details about the problems here.

twilco committed:
AX: The accessibility tree doesn't update when selection is cleared
https://bugs.webkit.org/show_bug.cgi?id=279484
rdar://135768164

Reviewed by Andres Gonzalez.

Prior to this patch, we never updated the isolated tree when selection
was cleared, only when it was changed from one node to another. This
meant that we would serve stale data for the AXSelectedTextMarkerRangeAttribute
API.

* LayoutTests/accessibility/mac/dynamic-selection-expected.txt: Added.
* LayoutTests/accessibility/mac/dynamic-selection.html: Added.
* LayoutTests/resources/accessibility-helper.js:
* Source/WebCore/editing/mac/FrameSelectionMac.mm:
(WebCore::FrameSelection::notifyAccessibilityForSelectionChange):

Canonical link: https://commits.webkit.org/283506@main
